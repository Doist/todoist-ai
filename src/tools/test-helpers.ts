import type { PersonalProject, Section, Task } from '@doist/todoist-api-typescript'

/**
 * Mapped task type matching the output of mapTask function.
 * Used for filter-based query test mocks.
 */
export type MappedTask = {
    id: string
    content: string
    description: string
    dueDate: string | undefined
    recurring: string | boolean
    priority: number
    projectId: string
    sectionId: string | null
    parentId: string | null
    labels: string[]
}

/**
 * Creates a mock Task with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockTask(overrides: Partial<Task> = {}): Task {
    return {
        id: '8485093748',
        content: 'Test task content',
        description: '',
        completedAt: null,
        labels: [],
        childOrder: 1,
        priority: 1,
        projectId: '6cfCcrrCFg2xP94Q',
        sectionId: null,
        parentId: null,
        url: 'https://todoist.com/showTask?id=8485093748',
        // Use correct property names from Task schema
        noteCount: 0,
        addedByUid: '713437',
        addedAt: '2025-08-13T22:09:56.123456Z',
        deadline: null,
        responsibleUid: null,
        assignedByUid: null,
        isCollapsed: false,
        isDeleted: false,
        duration: null,
        checked: false,
        updatedAt: '2025-08-13T22:09:56.123456Z',
        due: null,
        dayOrder: 0,
        userId: '713437',
        ...overrides,
    }
}

/**
 * Creates a mock Section with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockSection(overrides: Partial<Section> = {}): Section {
    return {
        id: 'section-123',
        projectId: '6cfCcrrCFg2xP94Q',
        sectionOrder: 1,
        userId: 'test-user',
        addedAt: '2024-01-01T00:00:00Z',
        updatedAt: '2024-01-01T00:00:00Z',
        archivedAt: null,
        isArchived: false,
        isDeleted: false,
        isCollapsed: false,
        name: 'Test Section',
        ...overrides,
    }
}

/**
 * Creates a mock PersonalProject with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockProject(overrides: Partial<PersonalProject> = {}): PersonalProject {
    return {
        id: '6cfCcrrCFg2xP94Q',
        name: 'Test Project',
        color: 'charcoal',
        isFavorite: false,
        isShared: false,
        parentId: null,
        inboxProject: false,
        viewStyle: 'list',
        url: 'https://todoist.com/projects/6cfCcrrCFg2xP94Q',
        isDeleted: false,
        updatedAt: '2025-08-13T22:09:55.841800Z',
        createdAt: '2025-08-13T22:09:55.841785Z',
        childOrder: 1,
        defaultOrder: 0,
        description: '',
        isCollapsed: false,
        canAssignTasks: false,
        isFrozen: false,
        isArchived: false,
        ...overrides,
    }
}

/**
 * Creates a mock API response object with results and nextCursor.
 */
export function createMockApiResponse<T>(
    results: T[],
    nextCursor: string | null = null,
): {
    results: T[]
    nextCursor: string | null
} {
    return {
        results,
        nextCursor,
    }
}

/**
 * Creates a simplified mapped task (matches mapTask output) for filter-based query tests.
 */
export function createMappedTask(overrides: Partial<MappedTask> = {}): MappedTask {
    return {
        id: TEST_IDS.TASK_1,
        content: 'Test task content',
        description: '',
        dueDate: undefined,
        recurring: false,
        priority: 1,
        projectId: TEST_IDS.PROJECT_TEST,
        sectionId: null,
        parentId: null,
        labels: [],
        ...overrides,
    }
}

/**
 * Common error messages used across tests.
 */
export const TEST_ERRORS = {
    API_RATE_LIMIT: 'API Error: Rate limit exceeded',
    API_UNAUTHORIZED: 'API Error: Unauthorized',
    INVALID_CURSOR: 'Invalid cursor format',
    INVALID_FILTER: 'Invalid filter query',
} as const

/**
 * Creates multiple test cases for parameterized testing.
 */
export function createTestCases<T, E = unknown>(
    cases: Array<{ name: string; input: T; expected?: E }>,
) {
    return cases
}

/**
 * Common mock IDs used across tests for consistency.
 */
export const TEST_IDS = {
    TASK_1: '8485093748',
    TASK_2: '8485093749',
    TASK_3: '8485093750',
    PROJECT_INBOX: 'inbox-project-id',
    PROJECT_WORK: 'work-project-id',
    PROJECT_TEST: '6cfCcrrCFg2xP94Q',
    SECTION_1: 'section-123',
    SECTION_2: 'section-456',
    USER_ID: '713437',
} as const
