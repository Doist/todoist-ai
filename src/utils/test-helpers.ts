import type { PersonalProject, Section, Task } from '@doist/todoist-api-typescript'
import { getToolOutput } from '../mcp-helpers'

/**
 * Mapped task type matching the output of mapTask function.
 * Used for filter-based query test mocks.
 */
export type MappedTask = {
    id: string
    content: string
    description: string
    dueDate: string | undefined
    recurring: string | boolean
    priority: number
    projectId: string
    sectionId: string | null
    parentId: string | null
    labels: string[]
    duration: string | null
    responsibleUid: string | null
    assignedByUid: string | null
}

/**
 * Creates a mock Task with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockTask(overrides: Partial<Task> = {}): Task {
    return {
        id: '8485093748',
        content: 'Test task content',
        description: '',
        completedAt: null,
        labels: [],
        childOrder: 1,
        priority: 1,
        projectId: '6cfCcrrCFg2xP94Q',
        sectionId: null,
        parentId: null,
        url: 'https://todoist.com/showTask?id=8485093748',
        // Use correct property names from Task schema
        noteCount: 0,
        addedByUid: '713437',
        addedAt: '2025-08-13T22:09:56.123456Z',
        deadline: null,
        responsibleUid: null,
        assignedByUid: null,
        isCollapsed: false,
        isDeleted: false,
        duration: null,
        checked: false,
        updatedAt: '2025-08-13T22:09:56.123456Z',
        due: null,
        dayOrder: 0,
        userId: '713437',
        ...overrides,
    }
}

/**
 * Creates a mock Section with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockSection(overrides: Partial<Section> = {}): Section {
    return {
        id: 'section-123',
        projectId: '6cfCcrrCFg2xP94Q',
        sectionOrder: 1,
        userId: 'test-user',
        addedAt: '2024-01-01T00:00:00Z',
        updatedAt: '2024-01-01T00:00:00Z',
        archivedAt: null,
        isArchived: false,
        isDeleted: false,
        isCollapsed: false,
        name: 'Test Section',
        url: 'https://todoist.com/sections/section-123',
        ...overrides,
    }
}

/**
 * Creates a mock PersonalProject with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockProject(overrides: Partial<PersonalProject> = {}): PersonalProject {
    return {
        id: '6cfCcrrCFg2xP94Q',
        name: 'Test Project',
        color: 'charcoal',
        isFavorite: false,
        isShared: false,
        parentId: null,
        inboxProject: false,
        viewStyle: 'list',
        url: 'https://todoist.com/projects/6cfCcrrCFg2xP94Q',
        isDeleted: false,
        updatedAt: '2025-08-13T22:09:55.841800Z',
        createdAt: '2025-08-13T22:09:55.841785Z',
        childOrder: 1,
        defaultOrder: 0,
        description: '',
        isCollapsed: false,
        canAssignTasks: false,
        isFrozen: false,
        isArchived: false,
        ...overrides,
    }
}

/**
 * Creates a mock API response object with results and nextCursor.
 */
export function createMockApiResponse<T>(
    results: T[],
    nextCursor: string | null = null,
): {
    results: T[]
    nextCursor: string | null
} {
    return {
        results,
        nextCursor,
    }
}

/**
 * Creates a simplified mapped task (matches mapTask output) for filter-based query tests.
 */
export function createMappedTask(overrides: Partial<MappedTask> = {}): MappedTask {
    return {
        id: TEST_IDS.TASK_1,
        content: 'Test task content',
        description: '',
        dueDate: undefined,
        recurring: false,
        priority: 1,
        projectId: TEST_IDS.PROJECT_TEST,
        sectionId: null,
        parentId: null,
        labels: [],
        duration: null,
        responsibleUid: null,
        assignedByUid: null,
        ...overrides,
    }
}

/**
 * Common error messages used across tests.
 */
export const TEST_ERRORS = {
    API_RATE_LIMIT: 'API Error: Rate limit exceeded',
    API_UNAUTHORIZED: 'API Error: Unauthorized',
    INVALID_CURSOR: 'Invalid cursor format',
    INVALID_FILTER: 'Invalid filter query',
} as const

/**
 * Creates multiple test cases for parameterized testing.
 */
export function createTestCases<T, E = unknown>(
    cases: Array<{ name: string; input: T; expected?: E }>,
) {
    return cases
}

/**
 * Extracts the text content from a tool output for snapshot testing.
 * This allows tests to match against just the text content while tools return structured output.
 */
export function extractTextContent(toolOutput: unknown): string {
    if (typeof toolOutput === 'string') {
        return toolOutput
    }

    if (typeof toolOutput === 'object' && toolOutput !== null && 'content' in toolOutput) {
        const output = toolOutput as { content: unknown }
        if (
            Array.isArray(output.content) &&
            output.content[0] &&
            typeof output.content[0] === 'object' &&
            output.content[0] !== null &&
            'type' in output.content[0] &&
            'text' in output.content[0] &&
            output.content[0].type === 'text'
        ) {
            return output.content[0].text as string
        }
    }

    throw new Error('Expected tool output to have text content')
}

/**
 * Extracts the structured content from a tool output for testing.
 * This handles both the new `structuredContent` field and legacy JSON-encoded content.
 */
export function extractStructuredContent(
    output: ReturnType<typeof getToolOutput>,
): Record<string, unknown> {
    // Check for new structuredContent field first
    if ('structuredContent' in output && typeof output.structuredContent === 'object') {
        return output.structuredContent as Record<string, unknown>
    }

    // Fall back to checking for JSON content in the content array
    if ('content' in output && Array.isArray(output.content)) {
        for (const item of output.content) {
            if (
                typeof item === 'object' &&
                item !== null &&
                'type' in item &&
                'text' in item &&
                item.type === 'text' &&
                item.mimeType === 'application/json'
            ) {
                return JSON.parse(item.text) as Record<string, unknown>
            }
        }
    }

    throw new Error('Expected tool output to have structured content')
}

/**
 * Common mock IDs used across tests for consistency.
 */
export const TEST_IDS = {
    TASK_1: '8485093748',
    TASK_2: '8485093749',
    TASK_3: '8485093750',
    PROJECT_INBOX: 'inbox-project-id',
    PROJECT_WORK: 'work-project-id',
    PROJECT_TEST: '6cfCcrrCFg2xP94Q',
    SECTION_1: 'section-123',
    SECTION_2: 'section-456',
    USER_ID: '713437',
} as const

/**
 * Fixed date for consistent test snapshots.
 * Use this instead of new Date() in tests to avoid snapshot drift.
 */
export const TODAY = '2025-08-17' as const
