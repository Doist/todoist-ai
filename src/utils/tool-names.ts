/**
 * Centralized tool names module
 *
 * This module provides a single source of truth for all tool names used throughout the codebase.
 * Each tool should import its own name from this module to avoid hardcoded strings.
 * This prevents outdated references when tool names change.
 */

export const TOOL_NAMES = {
    // Task management tools
    TASKS_ADD_MULTIPLE: 'tasks-add-multiple',
    TASKS_COMPLETE_MULTIPLE: 'tasks-complete-multiple',
    TASKS_UPDATE_MULTIPLE: 'tasks-update-multiple',
    TASKS_SEARCH: 'tasks-search',
    TASKS_LIST_BY_DATE: 'tasks-list-by-date',
    TASKS_LIST_COMPLETED: 'tasks-list-completed',
    TASKS_LIST_FOR_CONTAINER: 'tasks-list-for-container',

    // Project management tools
    PROJECTS_LIST: 'projects-list',
    PROJECTS_MANAGE: 'projects-manage',

    // Section management tools
    SECTIONS_SEARCH: 'sections-search',
    SECTIONS_MANAGE: 'sections-manage',

    // General tools
    OVERVIEW: 'overview',
    DELETE_ONE: 'delete-one',
} as const

// Type for all tool names
export type ToolName = (typeof TOOL_NAMES)[keyof typeof TOOL_NAMES]
