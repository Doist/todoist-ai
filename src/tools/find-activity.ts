import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { mapActivityEvent } from '../tool-helpers.js'
import { ApiLimits } from '../utils/constants.js'
import { summarizeList } from '../utils/response-builders.js'
import { ToolNames } from '../utils/tool-names.js'

const { FIND_TASKS, USER_INFO } = ToolNames

const ArgsSchema = {
    objectType: z
        .enum(['task', 'project', 'comment'])
        .optional()
        .describe('Type of object to filter by.'),

    objectId: z
        .string()
        .optional()
        .describe('Filter by specific object ID (task, project, or comment).'),

    eventType: z
        .enum([
            'added',
            'updated',
            'deleted',
            'completed',
            'uncompleted',
            'archived',
            'unarchived',
            'shared',
            'left',
        ])
        .optional()
        .describe('Type of event to filter by.'),

    projectId: z.string().optional().describe('Filter events by parent project ID.'),

    taskId: z.string().optional().describe('Filter events by parent task ID (for subtask events).'),

    initiatorId: z.string().optional().describe('Filter by the user ID who initiated the event.'),

    limit: z
        .number()
        .int()
        .min(1)
        .max(ApiLimits.ACTIVITY_MAX)
        .default(ApiLimits.ACTIVITY_DEFAULT)
        .describe('Maximum number of activity events to return.'),

    cursor: z
        .string()
        .optional()
        .describe('Pagination cursor for retrieving the next page of results.'),
}

const findActivity = {
    name: ToolNames.FIND_ACTIVITY,
    description:
        'Retrieve recent activity logs to monitor and audit changes in Todoist. Shows events from all users by default (use initiatorId to filter by specific user). Track task completions, updates, deletions, project changes, and more with flexible filtering. Note: Date-based filtering is not supported by the Todoist API.',
    parameters: ArgsSchema,
    async execute(args, client) {
        const { objectType, objectId, eventType, projectId, taskId, initiatorId, limit, cursor } =
            args

        // Build API arguments
        const apiArgs: Parameters<typeof client.getActivityLogs>[0] = {
            limit,
            cursor: cursor ?? null,
        }

        // Add optional filters
        if (objectType) apiArgs.objectType = objectType
        if (objectId && objectId !== 'remove') apiArgs.objectId = objectId
        if (eventType) apiArgs.eventType = eventType
        if (projectId) apiArgs.parentProjectId = projectId
        if (taskId) apiArgs.parentItemId = taskId
        if (initiatorId) apiArgs.initiatorId = initiatorId

        // Fetch activity logs from API
        const { results, nextCursor } = await client.getActivityLogs(apiArgs)
        const events = results.map(mapActivityEvent)

        // Generate text content
        const textContent = generateTextContent({
            events,
            args,
            nextCursor,
        })

        return getToolOutput({
            textContent,
            structuredContent: {
                events,
                nextCursor,
                totalCount: events.length,
                hasMore: Boolean(nextCursor),
                appliedFilters: args,
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

function generateTextContent({
    events,
    args,
    nextCursor,
}: {
    events: ReturnType<typeof mapActivityEvent>[]
    args: z.infer<z.ZodObject<typeof ArgsSchema>>
    nextCursor: string | null
}) {
    // Generate subject description
    let subject = 'Activity events'

    // Build subject based on filters
    const subjectParts: string[] = []
    if (args.eventType) {
        subjectParts.push(`${args.eventType}`)
    }
    if (args.objectType) {
        const objectLabel = args.objectType === 'task' ? 'tasks' : `${args.objectType}s`
        subjectParts.push(objectLabel)
    }

    if (subjectParts.length > 0) {
        subject = `Activity: ${subjectParts.join(' ')}`
    }

    // Generate filter hints
    const filterHints: string[] = []

    if (args.objectId) {
        filterHints.push(`object ID: ${args.objectId}`)
    }
    if (args.projectId) {
        filterHints.push(`project: ${args.projectId}`)
    }
    if (args.taskId) {
        filterHints.push(`task: ${args.taskId}`)
    }
    if (args.initiatorId) {
        filterHints.push(`initiator: ${args.initiatorId}`)
    }

    // Generate helpful suggestions for empty results
    const zeroReasonHints: string[] = []
    if (events.length === 0) {
        zeroReasonHints.push('No activity events match the specified filters')
        zeroReasonHints.push('Note: Activity logs only show recent events')

        if (args.eventType) {
            zeroReasonHints.push(`Try removing the eventType filter (${args.eventType})`)
        }
        if (args.objectType) {
            zeroReasonHints.push(`Try removing the objectType filter (${args.objectType})`)
        }
        if (args.objectId || args.projectId || args.taskId) {
            zeroReasonHints.push('Verify the object ID is correct')
        }
    }

    // Generate contextual next steps
    const nextSteps: string[] = []
    if (events.length > 0) {
        // Suggest related tools based on what was found
        const hasTaskEvents = events.some((e) => e.objectType === 'task' || e.objectType === 'item')
        const hasCompletions = events.some((e) => e.eventType === 'completed')

        if (hasTaskEvents) {
            nextSteps.push(`Use ${FIND_TASKS} to view current task details`)
        }
        if (hasCompletions) {
            nextSteps.push('Review completed tasks to track productivity')
        }
        if (args.initiatorId) {
            nextSteps.push(`Use ${USER_INFO} to get details about the user`)
        }

        // Suggest narrowing down if too many results
        if (events.length >= args.limit && !nextCursor) {
            nextSteps.push('Add more specific filters to narrow down results')
        }
    }

    return summarizeList({
        subject,
        count: events.length,
        limit: args.limit,
        nextCursor: nextCursor ?? undefined,
        filterHints,
        previewLines: previewActivityEvents(events, Math.min(events.length, args.limit)),
        zeroReasonHints,
        nextSteps,
    })
}

/**
 * Formats activity events into readable preview lines
 */
function previewActivityEvents(events: ReturnType<typeof mapActivityEvent>[], limit = 10): string {
    const previewEvents = events.slice(0, limit)
    const lines = previewEvents.map(formatActivityEventPreview).join('\n')

    // If we're showing fewer events than the total, add an indicator
    if (events.length > limit) {
        const remaining = events.length - limit
        return `${lines}\n    ... and ${remaining} more event${remaining === 1 ? '' : 's'}`
    }

    return lines
}

/**
 * Formats a single activity event into a readable preview line
 */
function formatActivityEventPreview(event: ReturnType<typeof mapActivityEvent>): string {
    const date = formatEventDate(event.eventDate)
    const eventLabel = `${event.eventType} ${event.objectType}`

    // Extract useful content from extraData if available
    let contentInfo = ''
    if (event.extraData) {
        const content =
            event.extraData.content || event.extraData.name || event.extraData.last_content
        if (content && typeof content === 'string') {
            // Truncate long content
            const truncated = content.length > 50 ? `${content.substring(0, 47)}...` : content
            contentInfo = ` • "${truncated}"`
        }
    }

    const objectId = event.objectId ? ` • id=${event.objectId}` : ''
    const initiator = event.initiatorId ? ` • by=${event.initiatorId}` : ' • system'
    const projectInfo = event.parentProjectId ? ` • project=${event.parentProjectId}` : ''

    return `    [${date}] ${eventLabel}${contentInfo}${objectId}${initiator}${projectInfo}`
}

/**
 * Formats an ISO date string to a more readable format
 */
function formatEventDate(isoDate: string): string {
    try {
        const date = new Date(isoDate)
        // Format as: Oct 23, 14:30 (in UTC for deterministic snapshots)
        const month = date.toLocaleDateString('en-US', { month: 'short', timeZone: 'UTC' })
        const day = date.toLocaleDateString('en-US', { day: 'numeric', timeZone: 'UTC' })
        const time = date.toLocaleTimeString('en-US', {
            hour: '2-digit',
            minute: '2-digit',
            hour12: false,
            timeZone: 'UTC',
        })
        return `${month} ${day}, ${time}`
    } catch {
        return isoDate
    }
}

export { findActivity }
