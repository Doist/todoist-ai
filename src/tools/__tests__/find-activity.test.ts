import type { ActivityEvent, TodoistApi } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import { extractTextContent } from '../../utils/test-helpers.js'
import { ToolNames } from '../../utils/tool-names.js'
import { findActivity } from '../find-activity.js'

// Mock the Todoist API
const mockTodoistApi = {
    getActivityLogs: jest.fn(),
} as unknown as jest.Mocked<TodoistApi>

const { FIND_ACTIVITY } = ToolNames

/**
 * Helper to create a mock activity event
 */
function createMockActivityEvent(overrides: Partial<ActivityEvent> = {}): ActivityEvent {
    return {
        id: 'event-123',
        objectType: 'task',
        objectId: 'task-456',
        eventType: 'added',
        eventDate: '2024-10-23T10:30:00Z',
        parentProjectId: 'project-789',
        parentItemId: null,
        initiatorId: 'user-001',
        extraData: null,
        ...overrides,
    }
}

describe(`${FIND_ACTIVITY} tool`, () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    describe('basic functionality', () => {
        it('should retrieve activity events with default parameters', async () => {
            const mockEvents: ActivityEvent[] = [
                createMockActivityEvent({
                    id: 'event-1',
                    eventType: 'added',
                    eventDate: '2024-10-23T10:00:00Z',
                }),
                createMockActivityEvent({
                    id: 'event-2',
                    eventType: 'completed',
                    eventDate: '2024-10-23T11:00:00Z',
                }),
            ]

            mockTodoistApi.getActivityLogs.mockResolvedValue({
                results: mockEvents,
                nextCursor: null,
            })

            const result = await findActivity.execute({ limit: 20 }, mockTodoistApi)

            expect(mockTodoistApi.getActivityLogs).toHaveBeenCalledWith({
                limit: 20,
                cursor: null,
            })

            expect(extractTextContent(result)).toMatchSnapshot()
        })

        it('should handle empty results', async () => {
            mockTodoistApi.getActivityLogs.mockResolvedValue({
                results: [],
                nextCursor: null,
            })

            const result = await findActivity.execute({ limit: 20 }, mockTodoistApi)

            expect(extractTextContent(result)).toMatchSnapshot()
        })

        it('should handle pagination with cursor', async () => {
            const mockEvents: ActivityEvent[] = Array.from({ length: 20 }, (_, i) =>
                createMockActivityEvent({
                    id: `event-${i}`,
                    objectId: `task-${i}`,
                }),
            )

            mockTodoistApi.getActivityLogs.mockResolvedValue({
                results: mockEvents,
                nextCursor: 'next-page-cursor',
            })

            const result = await findActivity.execute(
                { limit: 20, cursor: 'current-cursor' },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getActivityLogs).toHaveBeenCalledWith({
                limit: 20,
                cursor: 'current-cursor',
            })

            expect(extractTextContent(result)).toContain('Pass cursor')
            expect(extractTextContent(result)).toContain('next-page-cursor')
        })
    })

    describe('filtering', () => {
        it.each([
            ['task', 'added'],
            ['project', 'updated'],
            ['comment', 'deleted'],
        ])('should filter by object type: %s', async (objectType, eventType) => {
            const mockEvents: ActivityEvent[] = [
                createMockActivityEvent({
                    objectType: objectType as ActivityEvent['objectType'],
                    eventType: eventType as ActivityEvent['eventType'],
                }),
            ]

            mockTodoistApi.getActivityLogs.mockResolvedValue({
                results: mockEvents,
                nextCursor: null,
            })

            const result = await findActivity.execute(
                { objectType: objectType as 'task' | 'project' | 'comment', limit: 20 },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getActivityLogs).toHaveBeenCalledWith({
                objectType,
                limit: 20,
                cursor: null,
            })

            expect(extractTextContent(result)).toContain(objectType)
        })

        it.each([
            ['added', 'task-1'],
            ['completed', 'task-2'],
            ['updated', 'task-3'],
            ['deleted', 'task-4'],
        ])('should filter by event type: %s', async (eventType, objectId) => {
            const mockEvents: ActivityEvent[] = [
                createMockActivityEvent({
                    eventType: eventType as ActivityEvent['eventType'],
                    objectId,
                }),
            ]

            mockTodoistApi.getActivityLogs.mockResolvedValue({
                results: mockEvents,
                nextCursor: null,
            })

            const result = await findActivity.execute(
                {
                    eventType: eventType as
                        | 'added'
                        | 'updated'
                        | 'deleted'
                        | 'completed'
                        | 'uncompleted'
                        | 'archived'
                        | 'unarchived'
                        | 'shared'
                        | 'left',
                    limit: 20,
                },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getActivityLogs).toHaveBeenCalledWith({
                eventType,
                limit: 20,
                cursor: null,
            })

            expect(extractTextContent(result)).toContain(eventType)
        })

        it.each([
            ['objectId', 'task-123', { objectId: 'task-123' }],
            ['projectId', 'project-abc', { parentProjectId: 'project-abc' }],
            ['taskId', 'parent-task-789', { parentItemId: 'parent-task-789' }],
            ['initiatorId', 'user-alice', { initiatorId: 'user-alice' }],
        ])('should filter by %s', async (filterName, filterId, expectedApiCall) => {
            const mockEvents: ActivityEvent[] = [createMockActivityEvent()]

            mockTodoistApi.getActivityLogs.mockResolvedValue({
                results: mockEvents,
                nextCursor: null,
            })

            const args: Record<string, unknown> = { limit: 20 }
            args[filterName] = filterId

            await findActivity.execute(
                args as Parameters<typeof findActivity.execute>[0],
                mockTodoistApi,
            )

            expect(mockTodoistApi.getActivityLogs).toHaveBeenCalledWith({
                ...expectedApiCall,
                limit: 20,
                cursor: null,
            })
        })

        it('should support multiple filters simultaneously', async () => {
            const mockEvents: ActivityEvent[] = [
                createMockActivityEvent({
                    objectType: 'task',
                    eventType: 'completed',
                    parentProjectId: 'project-work',
                    initiatorId: 'user-bob',
                }),
            ]

            mockTodoistApi.getActivityLogs.mockResolvedValue({
                results: mockEvents,
                nextCursor: null,
            })

            const result = await findActivity.execute(
                {
                    objectType: 'task',
                    eventType: 'completed',
                    projectId: 'project-work',
                    initiatorId: 'user-bob',
                    limit: 50,
                },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getActivityLogs).toHaveBeenCalledWith({
                objectType: 'task',
                eventType: 'completed',
                parentProjectId: 'project-work',
                initiatorId: 'user-bob',
                limit: 50,
                cursor: null,
            })

            expect(extractTextContent(result)).toMatchSnapshot()
        })
    })

    describe('content extraction', () => {
        it('should extract task content from extraData', async () => {
            const mockEvents: ActivityEvent[] = [
                createMockActivityEvent({
                    eventType: 'added',
                    extraData: { content: 'Buy groceries' },
                }),
            ]

            mockTodoistApi.getActivityLogs.mockResolvedValue({
                results: mockEvents,
                nextCursor: null,
            })

            const result = await findActivity.execute({ limit: 20 }, mockTodoistApi)

            expect(extractTextContent(result)).toContain('Buy groceries')
        })

        it('should handle system-generated events with no initiator', async () => {
            const mockEvents: ActivityEvent[] = [
                createMockActivityEvent({
                    initiatorId: null,
                    eventType: 'completed',
                }),
            ]

            mockTodoistApi.getActivityLogs.mockResolvedValue({
                results: mockEvents,
                nextCursor: null,
            })

            const result = await findActivity.execute({ limit: 20 }, mockTodoistApi)

            expect(extractTextContent(result)).toContain('system')
        })

        it('should truncate long content', async () => {
            const longContent = 'A'.repeat(100)
            const mockEvents: ActivityEvent[] = [
                createMockActivityEvent({
                    extraData: { content: longContent },
                }),
            ]

            mockTodoistApi.getActivityLogs.mockResolvedValue({
                results: mockEvents,
                nextCursor: null,
            })

            const result = await findActivity.execute({ limit: 20 }, mockTodoistApi)

            expect(extractTextContent(result)).toContain('...')
            expect(extractTextContent(result)).not.toContain(longContent)
        })
    })
})
