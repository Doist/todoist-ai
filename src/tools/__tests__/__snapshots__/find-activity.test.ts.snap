// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`find-activity tool basic functionality should handle empty results 1`] = `
"Activity events: 0 (limit 20).
No results. No activity events match the specified filters; Note: Activity logs only show recent events."
`;

exports[`find-activity tool basic functionality should retrieve activity events with default parameters 1`] = `
"Activity events: 2 (limit 20).
Preview:
    [Oct 23, 10:00] added task • id=task-456 • by=user-001 • project=project-789
    [Oct 23, 11:00] completed task • id=task-456 • by=user-001 • project=project-789
Next:
- Use find-tasks to view current task details
- Review completed tasks to track productivity"
`;

exports[`find-activity tool filtering should support multiple filters simultaneously 1`] = `
"Activity: completed tasks: 1 (limit 50).
Filter: project: project-work; initiator: user-bob.
Preview:
    [Oct 23, 10:30] completed task • id=task-456 • by=user-bob • project=project-work
Next:
- Use find-tasks to view current task details
- Review completed tasks to track productivity
- Use user-info to get details about the user"
`;
